File: sdk/typescript/examples/simple-worker.ts
/**
 * Example: Simple task worker using the VELOCITY-WorkFlow TypeScript SDK.
 *
 * Demonstrates:
 *   - Connecting to the VELOCITY-WorkFlow server
 *   - Registering for a task queue
 *   - Polling for tasks in a loop
 *   - Executing task logic
 *   - Error handling
 *   - Graceful shutdown on SIGINT / SIGTERM
 *
 * Prerequisites:
 *   1. Start the VELOCITY-WorkFlow server:
 *      cd VELOCITY-WorkFlow/src/Velocity.Workflow.Server && dotnet run
 *
 *   2. Install dependencies:
 *      cd VELOCITY-WorkFlow/sdk/typescript && npm install
 *
 *   3. Run this worker:
 *      npx ts-node examples/simple-worker.ts
 */

import { VelocityClient, VelocityError } from '../src';

// ── Configuration ────────────────────────────────────────────────────────
const SERVER_ADDR = 'localhost:50051';
const TASK_QUEUE = 'orders';
const POLL_INTERVAL_MS = 1000;

// ── Graceful shutdown ────────────────────────────────────────────────────
let shutdownRequested = false;

process.on('SIGINT', () => {
  console.log('[worker] Received SIGINT — shutting down gracefully...');
  shutdownRequested = true;
});

process.on('SIGTERM', () => {
  console.log('[worker] Received SIGTERM — shutting down gracefully...');
  shutdownRequested = true;
});

// ── Task handlers ────────────────────────────────────────────────────────
interface Task {
  workflow_key: string;
  workflow_type: string;
  input: string;
}

type TaskHandler = (task: Task) => Promise<Record<string, unknown>>;

async function processOrder(task: Task): Promise<Record<string, unknown>> {
  const payload = JSON.parse(task.input || '{}');
  const orderId = payload.order_id ?? 'unknown';
  console.log(`[worker] Processing order ${orderId}`);
  // Simulate async work
  await new Promise((resolve) => setTimeout(resolve, 50));
  return { status: 'shipped', order_id: orderId };
}

const TASK_HANDLERS: Record<string, TaskHandler> = {
  'order-processing': processOrder,
};

// ── Worker loop ──────────────────────────────────────────────────────────
async function runWorker(): Promise<void> {
  console.log('[worker] Starting VELOCITY-WorkFlow TypeScript worker');
  console.log(`[worker] Server: ${SERVER_ADDR} | Queue: ${TASK_QUEUE}`);

  const client = new VelocityClient(SERVER_ADDR);

  try {
    const connected = await client.connect();
    if (!connected) throw new Error('Failed to connect to server');
    console.log(`[worker] Registered on task queue '${TASK_QUEUE}'`);

    while (!shutdownRequested) {
      try {
        const task: Task | null = await client.pollTask(TASK_QUEUE, 2000);

        if (task === null) {
          await sleep(POLL_INTERVAL_MS);
          continue;
        }

        const handler = TASK_HANDLERS[task.workflow_type];
        if (!handler) {
          console.warn(`[worker] No handler for task type '${task.workflow_type}' — skipping`);
          await client.failTask(task.workflow_key, `No handler for ${task.workflow_type}`);
          continue;
        }

        const result = await handler(task);
        await client.completeWorkflow(task.workflow_key, JSON.stringify(result));
        console.log(`[worker] Task '${task.workflow_type}' completed successfully`);
      } catch (err) {
        if (err instanceof VelocityError) {
          console.error(`[worker] Velocity error: ${err.message}`);
        } else {
          console.error(`[worker] Unexpected error:`, err);
        }
        await sleep(POLL_INTERVAL_MS);
      }
    }
  } catch (err) {
    console.error('[worker] Fatal error:', err);
  } finally {
    await client.close();
    console.log('[worker] Shut down cleanly');
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

runWorker().catch(console.error);

File: sdk/python/examples/simple_worker.py
#!/usr/bin/env python3
"""
Example: Simple task worker using the VELOCITY-WorkFlow Python SDK.

Demonstrates:
  - Connecting to the VELOCITY-WorkFlow server
  - Registering for a task queue
  - Polling for tasks in a loop
  - Executing task logic
  - Graceful error handling
  - Graceful shutdown on SIGINT / SIGTERM

Prerequisites:
  1. Start the VELOCITY-WorkFlow server:
     cd VELOCITY-WorkFlow/src/Velocity.Workflow.Server && dotnet run

  2. Install the SDK:
     cd VELOCITY-WorkFlow/sdk/python && pip install -r requirements.txt

  3. Run this worker:
     python examples/simple_worker.py
"""

import os
import signal
import sys
import time
import json
import logging

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from velocity_sdk import VelocityClient, WorkflowStatus, VelocityError

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(message)s",
)
logger = logging.getLogger("velocity-worker")

# ── Configuration ────────────────────────────────────────────────────────
SERVER_ADDR = "localhost:7234"
TASK_QUEUE = "orders"
POLL_INTERVAL_SEC = 1.0
MAX_RETRIES = 3

# ── Graceful shutdown flag ───────────────────────────────────────────────
shutdown_requested = False


def handle_signal(signum, frame):
    global shutdown_requested
    logger.info("Received signal %s — shutting down gracefully...", signum)
    shutdown_requested = True


signal.signal(signal.SIGINT, handle_signal)
signal.signal(signal.SIGTERM, handle_signal)


# ── Task handlers ────────────────────────────────────────────────────────
def process_order(task):
    """Process an incoming order task."""
    payload = json.loads(task.get("input", "{}"))
    order_id = payload.get("order_id", "unknown")
    logger.info("Processing order %s", order_id)
    # Simulate work
    time.sleep(0.05)
    return {"status": "shipped", "order_id": order_id}


TASK_HANDLERS = {
    "order-processing": process_order,
}


# ── Worker loop ──────────────────────────────────────────────────────────
def run_worker():
    logger.info("Starting VELOCITY-WorkFlow Python worker")
    logger.info("Server: %s | Queue: %s", SERVER_ADDR, TASK_QUEUE)

    client = VelocityClient(SERVER_ADDR)

    try:
        logger.info("Worker registered on task queue '%s'", TASK_QUEUE)

        while not shutdown_requested:
            try:
                # Poll for a workflow task from the server
                task = client.poll_task(TASK_QUEUE, timeout_ms=2000)

                if task is None:
                    logger.debug("No task available — retrying")
                    time.sleep(POLL_INTERVAL_SEC)
                    continue

                task_type = task.get("workflow_type", "unknown")
                handler = TASK_HANDLERS.get(task_type)

                if handler is None:
                    logger.warning("No handler for task type '%s' — skipping", task_type)
                    client.fail_task(task["workflow_key"], f"No handler for {task_type}")
                    continue

                # Execute the task
                result = handler(task)
                client.complete_workflow(task["workflow_key"], json.dumps(result).encode())
                logger.info("Task '%s' completed successfully", task_type)

            except VelocityError as exc:
                logger.error("Velocity error while processing task: %s", exc)
                time.sleep(POLL_INTERVAL_SEC)

            except Exception as exc:
                logger.exception("Unexpected error processing task: %s", exc)
                time.sleep(POLL_INTERVAL_SEC)

    except Exception as exc:
        logger.exception("Fatal worker error: %s", exc)
    finally:
        client.close()
        logger.info("Worker shut down cleanly")


if __name__ == "__main__":
    run_worker()
